packages/block/src/blockchainMachine.ts

```diff
@@ -1,27 +1,29 @@
 import { BlockProcessor } from "./blockProcessor";
 import { IBlockStub } from "./block";
-import { StartStopService, Lock } from "@pisa-research/utils";
-import { ApplicationError } from "@pisa-research/errors";
+import { StartStopService, Lock, logger } from "@pisa-research/utils";
+import { ConfigurationError, ArgumentError } from "@pisa-research/errors";
 import { Component, AnchorState, ComponentAction } from "./component";
 import { CachedKeyValueStore, ItemAndId } from "./cachedKeyValueStore";
 import { BlockItemStore } from "./blockItemStore";
 
 /**
- * Generic class to handle the anchor state of a blockchain state machine.
- *
- * For each block that is added to the cache (and for each added component), this will compute the new anchor state for
- * that component. Moreover, every time a "new head" event is emitted, this class will use the component's `detectChanges`
- * function to compute the appropriate actions, by comparing the newly computed anchor state with the anchor state of the
- * closest ancestor that was emitted as "new head". Since the latter might no longer be in the BlockCache, its anchor state
- * is propagated in each subsequent block; thus, every block stores its anchor state, and the "closest emitted ancestor"'s one.
+ * Blockchain machine functionality
  */
-export class BlockchainMachine<TBlock extends IBlockStub> extends StartStopService {
-    private components: Component<AnchorState, IBlockStub, ComponentAction>[] = [];
-    private componentNames: Set<string> = new Set();
-
+export class BlockchainMachine<TBlock extends IBlockStub> {
     // lock used to make sure that all events are processed in order
     private lock = new Lock();
 
+    constructor(
+        private readonly actionStore: CachedKeyValueStore<ComponentAction>,
+        public readonly blockItemStore: BlockItemStore<TBlock>,
+        private readonly components: Component<AnchorState, TBlock, ComponentAction>[]
+    ) {
+        const duplicateNames = components.map(c => c.name).filter((n, i) => components.map(c => c.name).lastIndexOf(n) !== i);
```

Comment:
This is O(n^2); I guess no big deal for now since we don't have too many components, but in principle it could be possible to have - say - tens of thousands of components; but I'd document the inefficiency with a comment.

Reply:
updated